ui/src/app/base/components/PowerTypeFields/PowerTypeFields.tsx

```diff
@@ -28,54 +31,121 @@ type Props = {
  * @param disabled - whether all fields should be disabled.
  * @param fieldScopes - the scopes of the fields to show.
  * @param powerParametersValueName - the power parameters "name" in the Formik form
+ * @param forConfiguration - whether the fields are being used for configuration of existing power parameters
+ * @param generateCert - whether a certificate should be generated (LXD-only)
+ * @param onCertRadioChange - function to run when changing certificate radio (LXD-only)
  * @returns list of Formik fields relevant to the chosen power type and field scopes.
  */
 const generateFields = (
   selectedPowerType: PowerType,
   disabled: boolean,
   fieldScopes: PowerFieldScope[],
-  powerParametersValueName: string
-) =>
-  selectedPowerType?.fields
-    ?.filter((field) => fieldScopes.includes(field.scope))
-    .map((field) => {
-      const { choices, field_type, label, name, required } = field;
-      return (
-        <FormikField
-          component={field_type === "choice" ? Select : Input}
-          disabled={disabled}
-          key={name}
-          label={label}
-          name={`${powerParametersValueName}.${name}`}
-          options={
-            field_type === "choice"
-              ? choices.map((choice) => ({
-                  key: `${name}-${choice[0]}`,
-                  label: choice[1],
-                  value: choice[0],
-                }))
-              : undefined
-          }
-          required={required}
-          type={
-            (field_type === "string" && "text") ||
-            (field_type === "password" && "password") ||
-            undefined
-          }
+  powerParametersValueName: string,
+  forConfiguration: boolean,
+  shouldGenerateCert: boolean,
+  onCertRadioChange: () => void
+) => {
+  const baseFields: ReactNode[] = [];
+  let customFields: ReactNode = null;
+  const fieldsInScope = selectedPowerType.fields.filter((field) =>
+    fieldScopes.includes(field.scope)
+  );
+  const isLxd = selectedPowerType.name === "lxd";
+
+  fieldsInScope.forEach((field) => {
+    // The authentication fields for the LXD power type are ignored here and
+    // special-cased below. All other power type fields are generated directly
+    // from the API data.
+    if (isLxd && ["certificate", "key", "password"].includes(field.name)) {
```

Comment:
If we are expecting these fields to be have the exact names "certificate", "key", "password" I think they should be an enum that is used everywhere so we are less likely to break this in the future.

Reply:
:+1: